File: src/topic_example/topic_talker/src/topic_talker.cc
/******************************************************************************
 * Copyright (c) 2022-2026 SEGAR. All Rights Reserved.
 * SPDX-License-Identifier: LicenseRef-Segar-Proprietary
 *
 * PROPRIETARY AND CONFIDENTIAL. See ./LICENSE
 * for license terms and restrictions.
 *****************************************************************************/

#include <memory>
#include <string>

#include "example/msg/String.hpp"

#include "segar/segar.h"

int main(int argc, char* argv[]) {
  RETURN_VAL_IF(!rti::segar::Init(argv[0]), EXIT_FAILURE);
  auto node = rti::segar::CreateNode("topic_talker");
  RETURN_VAL_IF(!node, EXIT_FAILURE);
  auto writer = node->CreateWriter<example::msg::String>("/topic/chatter");
  RETURN_VAL_IF(!writer, EXIT_FAILURE);
  uint32_t seq = 0;
  auto callback = [&writer, &seq]() {
    auto msg = std::make_shared<example::msg::String>();
    msg->data(std::to_string(seq++));
    AINFO_IF(!writer->Write(msg)) << "Failed to write msg:" << msg->data();
    AINFO << "Sent message: " << msg->data();
  };
  // 1hz
  auto timer = std::make_shared<rti::segar::Timer>(1000, callback, false);
  timer->Start();
  rti::segar::WaitForShutdown();
  return EXIT_SUCCESS;
}

File: src_python/topic_example/topic_talker/src/topic_talker.py
#!/usr/bin/env python3

import time

from example.msg import String
from segar.python import segar


def main():
    if not segar.Init("topic_talker"):
        return 1
    node = segar.CreateNode("topic_talker")
    if node is None:
        return 1
    writer = node.CreateWriter("/topic/chatter", String)
    if writer is None:
        print("[topic_talker] create writer failed")
        return 1

    seq = 0
    while segar.OK():
        msg = String(data=str(seq))
        if writer.Write(msg):
            print(f"[topic_talker] sent: {msg.data}")
        else:
            print(f"[topic_talker] write failed: {msg.data}")
        seq += 1
        time.sleep(1.0)

    segar.WaitForShutdown()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
